internal/graph/check.go

```diff
@@ -805,64 +829,151 @@ func buildTupleKeyConditionFilter(ctx context.Context, reqCtx *structpb.Struct,
+	pool.Go(func(ctx context.Context) error {
+		defer close(usersetsChan)
+		for {
+			t, err := iter.Next(ctx)
+			if err != nil {
+				if errors.Is(err, storage.ErrIteratorDone) {
+					break
+				}
+				// cancelled doesn't need to flush nor send errors back to main routine
+				if errors.Is(err, context.Canceled) {
+					return nil
+				}
+
+				telemetry.TraceError(span, err)
+				usersetsChan <- usersets{err: err}
+				return nil
+			}
+
+			objectRel, objectID, err := usersetDetails(t)
+			if err != nil {
+				usersetsChan <- usersets{err: err}
+			}
+
+			if _, ok := usersetsMap[objectRel]; !ok {
+				if len(usersetsMap) >= 1 {
+					// flush previous results
+					for k, v := range usersetsMap {
+						select {
+						case <-ctx.Done():
+							return nil
+						case usersetsChan <- usersets{
+							objectRelation: k,
+							objectIDs:      v,
+						}:
+							delete(usersetsMap, k)
+						}
+					}
+				}
+				usersetsMap[objectRel] = storage.NewSortedSet()
+			}
+
+			usersetsMap[objectRel].Add(objectID)
+			if usersetsMap[objectRel].Size() > 1000 {
+				select {
+				case <-ctx.Done():
+					return nil
+				case usersetsChan <- usersets{
+					objectRelation: objectRel,
+					objectIDs:      usersetsMap[objectRel],
+				}:
+					// flushed chunk, restarting objectRel entry
+					usersetsMap[objectRel] = storage.NewSortedSet()
+				}
 			}
-			telemetry.TraceError(span, err)
-			return nil, err
 		}
 
-		object, relation := tuple.SplitObjectRelation(t.GetUser())
-		objectType, objectID := tuple.SplitObject(object)
-		terminalRelations := typesys.GetTerminalRelations(objectType, relation, reqUserType)
-		// the terminalRelations is expected to be 1 (as we checked earlier in typesys.UsersetCanFastPath)
-		if len(terminalRelations) != 1 {
-			return nil, fmt.Errorf("expected exactly one terminal relation for fast path, received %d", len(terminalRelations))
+		// flush all remaining entries
+		for k, v := range usersetsMap {
+			select {
+			case <-ctx.Done():
+				return nil
+			case usersetsChan <- usersets{
+				objectRelation: k,
+				objectIDs:      v,
+			}:
+				// noop
+			}
 		}
-		computedRelation := terminalRelations[0]
-		objectRel := tuple.ToObjectRelationString(objectType, computedRelation)
-		if _, ok := usersetsMap[objectRel]; !ok {
-			usersetsMap[objectRel] = storage.NewSortedSet()
+
+		return nil
+	})
+	defer func() {
+		cancelFunc()
+		_ = pool.Wait() // error handled through the channel
+	}()
+	var finalErr error
+
+ConsumerLoop:
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case newBatch, channelOpen := <-usersetsChan:
+			if !channelOpen {
+				break ConsumerLoop
+			}
+			if newBatch.err != nil {
+				finalErr = newBatch.err
+				break ConsumerLoop
+			}
+			objectRel := newBatch.objectRelation
+			objectIDs := newBatch.objectIDs
+
+			resp, err := c.buildCheckAssociatedObjects(req, objectRel, objectIDs)(ctx)
+			if err != nil {
+				finalErr = err
```

Comment:
can you explain why do we break out of loop if newBatch.err != nil above but not breaking out of the loop here?